src/games/rcll/net.clp

```diff
@@ -197,6 +197,28 @@
   (pb-destroy ?attmsg)
 )
 
+(defrule send-delivery-msg
+  ?pd <- (product-delivered (id ?id) (team ?team) (order ?order)
+                            (game-time ?time) (confirmed FALSE)
+                            (delivery-gate ?gate))
+  =>
+  (bind ?msg (pb-create "llsf_msgs.UnconfirmedDelivery"))
```

Comment:
I'm still not convinced.

The sub-message can be more concise and only needs to carry the delivery time(s), maybe a "delivery id" (haven't checked the remaining code if this actually were still necessary, I suspect not, it is rather for this very mapping introduced by the separate message that it needs to exist). The rest of the information is already contained in the order itself.

> * It will not (easily) be possible to only send the UnconfirmedDelivery to the shell. We currently don't do that, but we may want to do that in the future.

Unless there is an actual use case this is just an unclear "maybe" statement.

> * We need to process the message separately anyway. So we would need to get the OrderInfo, and extract the UnconfirmedDeliveries. So we don't gain much by embedding it in the OrderInfo.

This is an inconvenience, but if the data still belongs logically together it should be sent together.

> * It separates the concerns. The purpose of the UnconfirmedDelivery is different to the OrderInfo: While the OrderInfo contains all orders and therefore is in a way part of the game state, the UnconfirmedDelivery is really a request by the refbox to the shell to confirm the delivery.

But an unconfirmed delivery is linked exactly to a specific order. So I see this as a property of an order, whether deliveries for a specific order have been claimed.

> * We can send the message immediately. While it doesn't make a big difference, this is cleaner by design. An event happens and we send the information to the shell. The AttentionMessage is treated the same way.

The AttentionMessage is special, since it needs immediate attention (if a referee needs to take care or evaluate something right away). For the unconfirmed delivery it doesn't matter (since we record the time of delivery it is no longer time critical). Machine state information, other order information, is sent every second or two. Sending the unconfirmed order along is just fine. No need for expedited submission to the shell. It eliminates yet another unnecessary code path.

So, considering the arguments, I would still suggest to fold this into an order. It just is a property of an order, not something by itself.

Reply:
> I'm still not convinced.
> 
> The sub-message can be more concise and only needs to carry the delivery time(s), maybe a "delivery id" (haven't checked the remaining code if this actually were still necessary, I suspect not, it is rather for this very mapping introduced by the separate message that it needs to exist). The rest of the information is already contained in the order itself.

No, the message needs to contain more than that. It needs the ID, because we can have multiple deliveries for one order, and we need to distinguish those when confirming the delivery. We either need the team or we need to have two lists in the order info, one for each team. We also need to have the gate because the order may allow any gate, but the delivery is always to one particular gate (although that most probably won't work with robocup-logistics/rcll-rulebook#21 anymore anyway).

> 
> > * We need to process the message separately anyway. So we would need to get the OrderInfo, and extract the UnconfirmedDeliveries. So we don't gain much by embedding it in the OrderInfo.
> 
> This is an inconvenience, but if the data still belongs logically together it should be sent together.
> 
> > * It separates the concerns. The purpose of the UnconfirmedDelivery is different to the OrderInfo: While the OrderInfo contains all orders and therefore is in a way part of the game state, the UnconfirmedDelivery is really a request by the refbox to the shell to confirm the delivery.
> 
> But an unconfirmed delivery is linked exactly to a specific order. So I see this as a property of an order, whether deliveries for a specific order have been claimed.

It's not tightly coupled to the order. A delivery is an event that needs to be reported to the shell to be approved or discarded. It belongs to an order, but it's not part of an order. As soon as it is confirmed, it will disappear. A delivery is also linked to exactly one team, this alone doesn't make it a property of the team. In the refbox, a delivery is represented by the `product-delivered` fact, which is not a slot of the order. In the shell, it's also not part of the order. It doesn't make sense to tie it to the order only for the messaging between refbox and shell, while keeping it separate from the order in both ends.